standalone/detectors/fasttrack/include/fasttrack.h

```diff
@@ -50,66 +59,372 @@ class Fasttrack : public Detector {
+    std::lock_guard<LockT> exLockT(g_lock);
+    auto it = happens_states.find(identifier);
+    if (it == happens_states.end()) {
+      it = create_happens(identifier);
+    }
+
+    ThreadState* thr = reinterpret_cast<ThreadState*>(tls);
+
+    thr->inc_vc();  // increment clock of thread and update happens state
+    it->second.update(thr->get_th_num(), thr->return_own_id());
+    // calls update(ThreadNum th_num, VC_EPOCH id)
+  }
+
+  void happens_after(tls_t tls, void* identifier) final {
+    {
+      std::lock_guard<LockT> exLockT(g_lock);
+      auto it = happens_states.find(identifier);
+      if (it == happens_states.end()) {
+        create_happens(identifier);
+        return;  // create -> no happens_before can be synced
+      } else {
+        ThreadState* thr = reinterpret_cast<ThreadState*>(tls);
+        // update vector clock of thread with happened before clocks
+        thr->update(it->second);  // calls update(VectorClock* other)
+      }
+    }
+  }
+
+  void allocate(tls_t tls, void* pc, void* addr, size_t size) final {
+#if REGARD_ALLOCS
+    size_t address = reinterpret_cast<size_t>(addr);
+    std::lock_guard<LockT> exLockT(g_lock);
+    if (allocs.find(address) == allocs.end()) {
+      create_alloc(address, size);
+    }
+#endif
+  }
+
+  // when we deallocate a variable we erase it from the VarState
+  void deallocate(tls_t tls, void* addr) final {
+#if REGARD_ALLOCS
+    size_t address = reinterpret_cast<size_t>(addr);
+    size_t end_addr;
+
+    std::lock_guard<LockT> exLockT(g_lock);
+    auto it = allocs.find(address);
+    if (it != allocs.end()) {
+      end_addr = address + it->second;
+    } else {
+      end_addr = 0;
+    }
+
+    while (address < end_addr) {  // we deallocate each address
+      /**
+       * \note Let's say we have a big vector that gets allocated. we track the
+       * allocation, but we don't create the Var state objects for the
+       * adresses. We create them 1 by 1 on each access. But when we deallocate
+       * we clear all of them
+       */
+
+      if (vars.find(address) != vars.end()) {
+        vars.erase(address);
+        address++;
+      } else {
+        address++;
+      }
+    }
+    allocs.erase(address);
+#endif
+  }
+
+  void detach(tls_t tls, tid_t thread_id) final {
+    /// \todo This is currently not supported
+    return;
+  }
+
+  void finish(tls_t tls, tid_t thread_id) final {
+    std::lock_guard<LockT> exLockT(g_lock);
+    /// just delete thread from list, no backward sync needed
+    VectorClock<>::ThreadNum th_num;
+    auto thrit = threads.find(thread_id);
+    th_num = thrit->second->get_th_num();
+    threads.erase(thrit);
+    cleanup(th_num);
+  }
+
+  void finalize() final {
+    std::lock_guard<LockT> lg1(g_lock);
+    vars.clear();
+    locks.clear();
+    happens_states.clear();
+    allocs.clear();
+    threads.clear();
+    shared_vcs.clear();
+    VectorClock<>::thread_ids.clear();
+
+    if (final_output) process_log_output();
```

Comment:
Please refer to the other comment pertaining to this file.
```suggestion
    if (log_flag) process_log_output();
```